ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/ConversationIndexMemory.java

```diff
@@ -0,0 +1,215 @@
+import static org.opensearch.ml.common.CommonValue.ML_MEMORY_META_INDEX;
+
+import java.util.Map;
+
+import org.opensearch.action.index.IndexRequest;
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.client.Client;
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.common.Strings;
+import org.opensearch.core.xcontent.ToXContent;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.index.query.BoolQueryBuilder;
+import org.opensearch.index.query.QueryBuilder;
+import org.opensearch.index.query.TermQueryBuilder;
+import org.opensearch.ml.common.spi.memory.Memory;
+import org.opensearch.ml.common.spi.memory.Message;
+import org.opensearch.ml.engine.indices.MLIndicesHandler;
+import org.opensearch.ml.memory.action.conversation.CreateConversationResponse;
+import org.opensearch.ml.memory.action.conversation.CreateInteractionResponse;
+import org.opensearch.search.builder.SearchSourceBuilder;
+import org.opensearch.search.sort.SortOrder;
+
+import lombok.Getter;
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2
+@Getter
+public class ConversationIndexMemory implements Memory {
+    public static final String TYPE = "conversation_index";
+    public static final String CONVERSATION_ID = "conversation_id";
+    public static final String FINAL_ANSWER = "final_answer";
+    public static final String CREATED_TIME = "created_time";
+    public static final String MEMORY_NAME = "memory_name";
+    public static final String MEMORY_ID = "memory_id";
+    public static final String APP_TYPE = "app_type";
+    public static int LAST_N_INTERACTIONS = 10;
+    protected String memoryMetaIndexName;
+    protected String memoryMessageIndexName;
+    protected String conversationId;
+    protected boolean retrieveFinalAnswer = true;
+    protected final Client client;
+    private final MLIndicesHandler mlIndicesHandler;
+    private MLMemoryManager memoryManager;
+
+    public ConversationIndexMemory(
+        Client client,
+        MLIndicesHandler mlIndicesHandler,
+        String memoryMetaIndexName,
+        String memoryMessageIndexName,
+        String conversationId,
+        MLMemoryManager memoryManager
+    ) {
+        this.client = client;
+        this.mlIndicesHandler = mlIndicesHandler;
+        this.memoryMetaIndexName = memoryMetaIndexName;
+        this.memoryMessageIndexName = memoryMessageIndexName;
+        this.conversationId = conversationId;
+        this.memoryManager = memoryManager;
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    @Override
+    public void save(String id, Message message) {
+        this.save(id, message, ActionListener.wrap(r -> { log.info("saved message into {} memory, session id: {}", TYPE, id); }, e -> {
+            log.error("Failed to save message to memory", e);
+        }));
```

Comment:
this method looks like a synchronized invocation, can we have a future returned instead of void?